homeassistant/components/light/mqtt_json.py

```diff
@@ -86,14 +88,25 @@
 
 async def async_setup_platform(hass: HomeAssistantType, config: ConfigType,
                                async_add_entities, discovery_info=None):
-    """Set up a MQTT JSON Light."""
-    if discovery_info is not None:
-        config = PLATFORM_SCHEMA(discovery_info)
+    """Set up MQTT JSON Light through configuration.yaml."""
+    await _async_setup_entity(hass, config, async_add_entities)
 
-    discovery_hash = None
-    if discovery_info is not None and ATTR_DISCOVERY_HASH in discovery_info:
-        discovery_hash = discovery_info[ATTR_DISCOVERY_HASH]
 
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up MQTT JSON Light dynamically through MQTT discovery."""
+    async def async_discover(discovery_payload):
+        """Discover and add a MQTT JSON Light."""
+        config = PLATFORM_SCHEMA(discovery_payload)
+        await _async_setup_entity(hass, config, async_add_entities,
+                                  discovery_payload[ATTR_DISCOVERY_HASH])
+    async_dispatcher_connect(
+        hass, MQTT_DISCOVERY_NEW.format(light.DOMAIN, DOMAIN),
+        async_discover)
+
+
+async def _async_setup_entity(hass, config, async_add_entities,
+                              discovery_hash=None):
+    """Set up the MQTT JSON Light."""
```

Comment:
Have you tested this? I believe the setup entry will never be called because the config entry is forwarded to the platform with the same name as the component (`mqtt`). That's a limitation of the config entry system and we should not change the config entry core just to work around this quirk.

There's been some discussion before (can't find the link rn), but the conclusion was that the `light.mqtt` platform would need to redirect the entity setup to mqtt_json if the platform doesn't match.

Reply:
I've only tested the unit tests, I'll do some "real" testing too.